Upgrade confluent-kafka-python to 1.6.1

aericson opened the pull request:
**Background**
Routine upgrade of confluent-kafka-python. The idea for the bump here came from a discussion on building a python 3.9 image on our https://github.com/fyndiq/python-confluent-kafka image. This is not necessary but seems reasonable to do both at once.

**Changes**
- Upgrade confluent-kafka-python to 1.6.1
- Remove avro-python3 and fastavro from requirements.in and setup.py and instead use `confluent-kafka-python[avro]`.
    - The advantage here is that we don't need to track the version constraint on those dependencies and instead let confluent-kafka-python handle it (https://github.com/confluentinc/confluent-kafka-python/blob/master/setup.py#L28-L33).

Could use a second opinion on the later, please let me know if you have any concerns.

Commit:
Upgrade confluent-kafka-python to 1.6.1

aericson commented:
> > * Upgrade confluent-kafka-python to 1.6.1
> > * Remove avro-python3 and fastavro from requirements.in and setup.py and instead use `confluent-kafka-python[avro]`.
> > 
> > Could use a second opinion on the later, please let me know if you have any concerns.
> 
> Could you please explain more in the description, why are we doing these changes?

Done! let me know if you need more info.

pythops reviewed (approved):
GJ 👌🏿

simonwahlgren commented:
Nice initiative! However, there are some things to keep in mind when bumping `confluent-kafka-python`:

1. Read all release notes from the current version to the version we want to upgrade to, for both `confluent-kafka-python` and `librdkafka`, in this case the following ones:
    - https://github.com/confluentinc/confluent-kafka-python/releases/tag/v1.6.0
    - https://github.com/confluentinc/confluent-kafka-python/releases/tag/v1.6.1
    - https://github.com/edenhill/librdkafka/releases/tag/v1.6.0
    - https://github.com/edenhill/librdkafka/releases/tag/v1.6.1

    Are there any new fixes/features we have to consider? Usually there are new/deprecated configurations between the minor versions at least. Some examples:

    #### confluent-kafka-python 1.6.0
    > Consumer destructor will no longer trigger consumer_close(),
consumer.close() must now be explicitly called if the application
wants to leave the consumer group properly and commit final offsets.

    This could potentially create issues for us if we didn't manually close the consumer, in this case it's already handled here: https://github.com/fyndiq/confluent_kafka_helpers/blob/master/confluent_kafka_helpers/consumer.py#L114

    #### confluent-kafka-python 1.6.1
    > Add **kwargs to legacy AvroProducer and AvroConsumer constructors to
support all Consumer and Producer base class constructor arguments, such
as logger
    
    Means we finally can use the same logging configuration as for the application. I would probably just fix this, but at least create a ticket for it.

    #### librdkafka 1.6.0
    > Sticky producer partitioning (sticky.partitioning.linger.ms) is
enabled by default (10 milliseconds) which affects the distribution of
randomly partitioned messages, where previously these messages would be
evenly distributed over the available partitions they are now partitioned
to a single partition for the duration of the sticky time
(10 milliseconds by default) before a new random sticky partition
is selected.

    Is this something we want (probably yes)?

    > KIP-447 Producer scalability for exactly once semantics - allows a single transactional producer to be used for multiple input partitions. Requires Apache Kafka 2.5 or later.
    
    Probably something we should look into once we upgrade to 2.5. Something I would create a ticket for at least so we don't forget about it.

    My point is that usually there are some deprecation and or new features/configurations we have to consider or implement. If we blindly just bump the library some things might break, or we might miss out on new nice features (idempotent producer). So we have to put some time into investigating the new changes at least.

2. The reason we pinned the Avro requirements before was because `fastavro` never was pinned in `confluent-kafka-python` (still not pinned). Historically this has caused some issues for us due to bugs in `fastavro` which resulted in broken messages due to encoding/decoding issues. Since we used event sourcing in many of our flows before this could (and did) create a lot of problems for us. So before when we (me) bumped `fastavro` we (I) had to make sure to test it a lot before rolling it out.

    However, since we don't use event sourcing any more one could argue to be a bit more relaxed here, I'm all for testing it out at least, in worst case we'll get some borked messages... :sweat_smile: 


Also, this might be a good opportunity to start using pre releases in PyPi so we can test stuff before creating an official release. Historically I would just install the lib from master in a less "important "service (`article-api-service`, `article-handler-service` etc.) and let it run for a while (a week) before I felt confident everything was working, then I would create a PyPi release and start rolling it out gradually. What do you think?

aericson commented:
Great points! I only had a quick look for breaking changes since it was a minor release. But should have looked more into it.

Hope I've addressed all the main ones now.

1) Support for passing logger to consumer/provider #65 also created a follow up issue in https://github.com/fyndiq/eventsourcing_helpers/issues/72
2) Pinning avro dependencies back #64 and bumped one version of fastavro to support python 3.9
3) "Look into KIP-447 when upgrading kafka 2.5" created on internal Jira, issue 2838, and updated idempotence config in #66

> Also, this might be a good opportunity to start using pre releases in PyPi so we can test stuff before creating an official release. Historically I would just install the lib from master in a less "important "service (article-api-service, article-handler-service etc.) and let it run for a while (a week) before I felt confident everything was working, then I would create a PyPi release and start rolling it out gradually. What do you think?

I think it's a good idea if we want to create public pre-releases. For us to test internally I think it's not worth the trouble and maybe it's better to just pin to a commit on master instead, such as:

`git+https://github.com/fyndiq/confluent_kafka_helpers.git@<commit>`

What do you think?

simonwahlgren commented:
> Great points! I only had a quick look for breaking changes since it was a minor release. But should have looked more into it.
> 
> Hope I've addressed all the main ones now.
> 
> 1. Support for passing logger to consumer/provider #65 also created a follow up issue in [fyndiq/eventsourcing_helpers#72](https://github.com/fyndiq/eventsourcing_helpers/issues/72)
> 2. Pinning avro dependencies back #64 and bumped one version of fastavro to support python 3.9
> 3. "Look into KIP-447 when upgrading kafka 2.5" created on internal Jira, issue 2838, and updated idempotence config in #66
> 
> > Also, this might be a good opportunity to start using pre releases in PyPi so we can test stuff before creating an official release. Historically I would just install the lib from master in a less "important "service (article-api-service, article-handler-service etc.) and let it run for a while (a week) before I felt confident everything was working, then I would create a PyPi release and start rolling it out gradually. What do you think?
> 
> I think it's a good idea if we want to create public pre-releases. For us to test internally I think it's not worth the trouble and maybe it's better to just pin to a commit on master instead, such as:
> 
> `git+https://github.com/fyndiq/confluent_kafka_helpers.git@<commit>`
> 
> What do you think?

Awesome! I totally agree that it's better to pin a commit hash instead of using master, then we get rid of the problem where the library might not get updated (even if there are new remote commits) if there are no changes to that layer in the Dockerfile.
